Sources/URLRouting/Fragment.swift

```diff
@@ -0,0 +1,74 @@
+import Parsing
+
+/// Parses a request's fragment subcomponent.
+public struct Fragment<FragmentParser: Parser>: Parser where FragmentParser.Input == Substring {
+    
+  @usableFromInline
+  let fragmentParser: FragmentParser
+  
+  @inlinable
+  public init(@ParserBuilder build: () -> FragmentParser) {
+    self.fragmentParser = build()
+  }
+  
+  /// Initializes a fragment parser from a substring conversion.
+  ///
+  /// Useful for parsing a request fragment in its entirety, for example as a String.
+  ///
+  /// ```swift
+  /// Fragment(.string)
+  /// ```
+  ///
+  /// - Parameter fragmentConversion: A conversion that transforms fragment into some other type.
+  @inlinable
+  public init<C>(_ fragmentConversion: C)
+  where FragmentParser == Parsers.MapConversion<Parsers.ReplaceError<Rest<FragmentParser.Input>>, C> {
+    self.fragmentParser = Rest().replaceError(with: .init()).map(fragmentConversion)
+  }
+  
+  /// Initializes a fragment parser from a substring conversion and a parser of the conversion's output.
+  ///
+  /// Useful for transforming a request fragment in its entirety, and then parsing on the transformed value.
+  ///
+  /// For example, if your server responds with a query string in the URL's fragment, you could create a
+  /// "fields" conversion that transforms a `Substring` into a `URLRequestData.Fields` type. With
+  /// this in place you could parse the individual ``Field`` elements in the following query-like URL
+  /// fragment:
+  ///
+  /// `/search#query=Blob`
+  ///
+  /// ```swift
+  /// Fragment(.fields) {
+  ///   Field("query", default: "")
+  /// }
+  /// ```
+  ///
+  /// - Parameters:
+  ///   - fragmentConversion: A conversion that transforms the fragment into some other type.
+  ///   - parsers: A parser builder that can parse the conversion's output.
+  @inlinable
+  public init<C: Conversion, P: Parser>(
+    _ fragmentConversion: C,
+    @ParserBuilder with parsers: () -> P
+  )
+  where FragmentParser == Parsers.Pipe<Parsers.MapConversion<Parsers.ReplaceError<Rest<Substring>>, C>, P> {
```

Comment:
I think it makes sense to have an alternate interface that takes a parser builder (just like `Body` does), but I think we should limit the number of "mapping" operators like this to higher-level constructs like `Route`, `Parse`, and `ParsePrint`, especially since there are more overloads to consider.

Effectively, your example could be something like:

```swift
ParsePrint(
  .fields {
    Field("query", default: "")
  }
) {
  Fragment()
}

// or:

Fragment().map(.fields {
  Field("query", default: "")
})
```

Reply:
I agree. I also don't think my use of `Pipe` here is correct. Is the above based on existing API or could you nudge me in the right direction as to how to achieve the example you suggested? After spending some time trying to work this out, I was unable to come up with anything successful.

BTW, here is the fields conversion I have locally. 

<details>
  <summary>StringToFields.swift</summary>
  
  ```swift
import Foundation
    import Parsing

    extension URLRequestData.Fields: _EmptyInitializable {
      public init() {
        self.fields = [:]
        self.isNameCaseSensitive = false
      }
    }

    extension Conversion where Self == Conversions.StringToFields {
      
      /// A conversion from `Substring` to `URLRequestData.Fields`.
      @inlinable
      public static var fields: Self { .init() }
      
      /// A conversion from `Substring` to `URLRequestData.Fields`.
      @inlinable
      public static func fields(
        isNameCaseSensitive: Bool = Self.defaultCaseSensitivity,
        sortedBy fieldSorting: @escaping Self.FieldSorting = Self.defaultFieldSorting
      ) -> Self { .init(isNameCaseSensitive: true, sortedBy: fieldSorting) }
    }

    extension Conversion where Output == Substring {
      /// Transforms this conversion to `String` into a conversion to `URLRequestData.Fields`.
      ///
      /// A fluent version of ``Conversion/fields(isNameCaseSensitive:sortedBy:)``.
      @inlinable
      public func fields(
        isNameCaseSensitive: Bool = Conversions.StringToFields.defaultCaseSensitivity,
        sortedBy fieldSorting: @escaping Conversions.StringToFields.FieldSorting = Conversions.StringToFields.defaultFieldSorting
      ) -> Conversions.Map<Self, Conversions.StringToFields> {
        
        self.map(.fields(isNameCaseSensitive: isNameCaseSensitive, sortedBy: fieldSorting))
        
      }
    }

    extension Conversions {
      /// A conversion from a Substring to URLRequestData.Fields.
      ///
      /// You will not typically need to interact with this type directly. Instead you will usually use
      /// the ``Conversion/fields(isNameCaseSensitive:sortedBy:)`` operation,
      /// which constructs this type.
      public struct StringToFields: Conversion {
        
        public typealias FieldSorting = (URLRequestData.Fields.Element, URLRequestData.Fields.Element) -> Bool
        
        public static let defaultCaseSensitivity = true
        public static let defaultFieldSorting: FieldSorting = { $0.key < $1.key }
        
        public var isNameCaseSensitive: Bool
        public var fieldSorting: FieldSorting
        
        @inlinable
        public init(
          isNameCaseSensitive: Bool = Self.defaultCaseSensitivity,
          sortedBy fieldSorting: @escaping FieldSorting = Self.defaultFieldSorting
        ) {
          self.isNameCaseSensitive = isNameCaseSensitive
          self.fieldSorting = fieldSorting
        }
        
        @inlinable
        public func apply(_ input: Substring) -> URLRequestData.Fields {
          var components = URLComponents()
          components.query = String(input)
          
          let query: [String : [String?]] = components.queryItems?
            .reduce(into: [:]) { query, item in
              query[item.name, default: []].append(item.value)
            } ?? [:]
          
          return .init(
            query.mapValues { $0.map { $0?[...] }[...] },
            isNameCaseSensitive: self.isNameCaseSensitive
          )
        }
        
        @inlinable
        public func unapply(_ output: URLRequestData.Fields) -> Substring {
          var components = URLComponents()
          components.queryItems = output
            .sorted(by: fieldSorting)
            .flatMap { name, values in
              values.map { URLQueryItem(name: name, value: $0.map(String.init)) }
            }
          return Substring(components.query ?? "")
        }
      }
    }
  ```
</details>

Would I just add a `@ParserBuilder` to this in order to get syntax like you suggested?

```swift
.fields {
  Field("query", default: "")
}
```

Reply:
I think so! If you need any more help working out the details, want to move this to a discussion since it's unrelated to the PR?

Reply:
Updated PR to remove my odd init method that performed the mapping. Will explore an alternative in my custom `fields` conversion.